File: 027.rb
# Considering quadratics of the form: n² + an + b, where |a| < 1000 
# and |b| < 1000, find the product of the coefficients, a and b, 
# for the quadratic expression that produces the maximum number of 
# primes for consecutive values of n, starting with n = 0.

require 'prime'

def calculate_max_consecutive_primes(a, b, primes)
    consecutives = 0
    non_prime_found = false
    n = 0
    while (not non_prime_found)
        if primes.include?(n**2 + a*n + b)
            consecutives += 1
        else
            return consecutives
        end
        n += 1
    end
end

primes = Prime.first(1000)
result = max_consecutives = 0
(-999..999).each do |a|
    (-999..999).each do |b|
        consecutives = calculate_max_consecutive_primes(a, b, primes)
        if consecutives > max_consecutives
            max_consecutives = consecutives
            result = a*b
        end
    end
end
puts result

File: clojure/src/project_euler_clojure/027.clj
;; Considering quadratics of the form: n² + an + b, where |a| < 1000
;; and |b| < 1000, find the product of the coefficients, a and b,
;; for the quadratic expression that produces the maximum number of
;; primes for consecutive values of n, starting with n = 0.

(ns project-euler-clojure.027)

(use 'clojure.set)

(def primes
  (loop [primes [] candidates (apply sorted-set (range 2 (inc 10000)))]
    (cond
      (empty? candidates) primes
      :else (let [new-prime (first candidates)
                  new-primes (conj primes new-prime)
                  multiples (range new-prime (inc 10000) new-prime)
                  new-candidates (difference candidates multiples)]
              (recur new-primes new-candidates)))))

(defn prime? [n]
  (not= -1 (.indexOf primes n)))

(def is-prime?-memo (memoize prime?))

;; at n = 0, n^2+ an + b = b, b = b. Therefore, b needs to be prime
(defn possible-solution-pairs []
  (for [a (range -999 1000)
        b (range -999 1000)
        :when (prime? b)]
    [a b]))

(defn quadratic-formula [a b n]
  (+ (+ (* n n)
        (* a n))
     b))

(defn max-primes [pair]
  (->> (iterate inc 0)
       (map #(quadratic-formula (first pair) (second pair) %))
       (take-while is-prime?-memo)
       (count)))

(defn problem-027 []
  (->> (possible-solution-pairs)
       (map #(hash-map :pair % :primes (max-primes %)))
       (sort-by :primes >)
       (first)
       (#(get % :pair))
       (reduce *)))
